test_volume.py

```diff
@@ -178,3 +179,198 @@ def test_maximum_number_of_volumes(server, create_volume):
+    """
+
+    # Attach and format volume to server
+    volume.attach(server)
+    time.sleep(5)
+    server.assert_run('sudo mkfs.ext4 /dev/sdb')
+    server.assert_run('sudo mount /dev/sdb /mnt')
+
+    # Create two files. The first is synced to disk with fsync, the second
+    # is not synced. Data might still be in-flight.
+    server.assert_run(
+        'sudo dd if=/dev/zero of=/mnt/synced count=1 bs=1M conv=fsync')
+    server.assert_run('sudo dd if=/dev/zero of=/mnt/not-synced count=1 bs=1M')
+
+    # Create snapshot
+    snapshot = volume.snapshot('snap')
+
+    # Write test file to volume
+    server.assert_run('sudo touch /mnt/after-snapshot')
+
+    # Try reverting while attached (should fail)
+    with pytest.raises(HTTPError) as error:
+        volume.revert(snapshot)
+
+    # Assert a HTTP 400 BadRequest response with a specific error message
+    assert error.value.response.status_code == 400
+    assert error.value.response.json()['detail'] == (
+        'Cannot revert non-root volumes while they are attached to a server.'
+    )
+
+    # Detach volume
+    server.assert_run('sudo umount /mnt')
+    volume.detach()
+    time.sleep(5)
+
+    # Revert volume to snapshot
+    volume.revert(snapshot)
+
+    # Reattach and mount the volume
+    volume.attach(server)
+    time.sleep(5)
+    server.assert_run('sudo mount /dev/sdb /mnt')
+
+    # Verify the test files are in the correct state
+    assert server.file_path_exists('/mnt/synced')
+    assert not server.file_path_exists('/mnt/after-snapshot')
+
+    # "Warn" if the file "not-synced" exists. This is not a failure because
+    # depending on the exact timing the data might get written to disk.
+    if server.file_path_exists('/mnt/not-synced'):
+        warn(
+            'File "not-synced" is included in the snapshot although it was '
+            'not explicitly synced.',
+        )
+
+
+def test_snapshot_volume_detached(server, volume):
+    """ Detached volumes can be snapshotted and reverted.
+
+    It is possible to create a snapshot of a volume which is detached
+    and to revert the volume back to this state.
+
+    Snapshots taken while the volume is detached and the filesystem unmounted
+    are always fully consistent. All data is written to the disk.
+    """
+
+    # Attach and format volume to server
+    volume.attach(server)
+    time.sleep(5)
+    server.assert_run('sudo mkfs.ext4 /dev/sdb')
+    server.assert_run('sudo mount /dev/sdb /mnt')
+
+    # Create two files. The first is synced to disk with fsync, the second
+    # is not synced. Data might still be in-flight.
+    server.assert_run(
+        'sudo dd if=/dev/zero of=/mnt/synced count=1 bs=1M conv=fsync')
+    server.assert_run('sudo dd if=/dev/zero of=/mnt/not-synced count=1 bs=1M')
+
+    # Unmount the volume
+    server.assert_run('sudo umount /mnt')
+
+    # Record the checksum of the first 1GiB of the volume
+    # (Checksumming the whole volume would take too much time.)
```

Comment:
Is this because we use sha256? Because if possible I think it would be nice to checksum the whole volume. Maybe use md5? How long does it really take?

Reply:
No this is because the `volume` fixture creates volumes of size 100G. While we could create a smaller SSD volume, for bulk volumes this is the minimal size.

I did not test another hash function, but with `sha256sum` it took over 10 minutes.

I'm not sure what the limiting factor is. I suspect virtual disk bandwidth even though the volume is thin allocated and no actual reads should be done for most of the volume.

Because of the thin allocation we would also mostly checksum zeros.

I also think that this can only be broken if some very low level Ceph function is broken. There should be enough testing around this in Ceph.